### PR TITLE
Extract UpdateConfigmap into e2e framework

### PR DESCRIPTION
**Pull request by zrdaley**
This function is required in several e2e tests. Pulling into the framework.go file will reduce code clones.

### CODE REVIEW

File: test/e2e/framework/framework.go [hunk 1]
```diff
@@ -279,3 +279,26 @@ func UpdateDeployment(kubeClientSet kubernetes.Interface, namespace string, name
 
 	return nil
 }
+
+// UpdateConfigmap updates the ingress-nginx config
```
Comment (andrewlouis93) [inline test/e2e/framework/framework.go]: Could we make this `updateNginxConfigMap`? So we have a sense of whose/which ConfigMap is updated?

File: test/e2e/framework/framework.go [hunk 2]
```diff
@@ -279,3 +279,26 @@ func UpdateDeployment(kubeClientSet kubernetes.Interface, namespace string, name
...
+	if config.Data == nil {
+		config.Data = map[string]string{}
+	}
+	config.Data[k] = v
+	_, err = c.CoreV1().ConfigMaps("ingress-nginx").Update(config)
+	if err != nil {
+		return err
+	}
+	time.Sleep(1 * time.Second)
+	return nil
```
Comment (andrewlouis93) [inline test/e2e/framework/framework.go]: Could we return a deep copy of the old configMap before the update? This way we can reset the configMap at the end of a test without mutating state

File: test/e2e/framework/framework.go (hunk 1 above)
Comment (andrewlouis93) [reply test/e2e/framework/framework.go]: Nice, but if we pass in the name of the configMap, it'd still be _somewhat_ specific since the method needs to know that we're talking about the namespace `ingress-nginx`. 

I think I'll go with `updateNginxConfigMap`, and if we find out that we need a generic one for the temporary deployments maybe - we can update :)

File: test/e2e/framework/framework.go (hunk 2 above)
Comment (zrdaley) [reply test/e2e/framework/framework.go]: Is this pattern used elsewhere in the framework? Returning the previous state from a function that is altering the config map seems a little bit confusing. I think it might be more logical to have a ConfigMapReset function

Comment (andrewlouis93) [reply test/e2e/framework/framework.go]: Hmm, I think you're right - it would be cleaner, but my concern is - how would the method know what is the default? The developer might have kubectl-d her own configMap, and I'm not sure assuming that [deploy/configmap.yaml](https://github.com/kubernetes/ingress-nginx/blob/master/test/e2e/wait-for-nginx.sh#L22) has the defaults is the right approach. 

Let me know what you think / if that resonates. When the e2e-tests run, ingress-nginx and everything in it should already be a thing, so I'm not sure at what point we'd capture the initial state

Comment (andrewlouis93) [reply test/e2e/framework/framework.go]: > Is this pattern used elsewhere in the framework

Not really, since there aren't any tests that require changing ingress-nginx. For example wherever `updateDeployment` is used - it doesn't need to reset anything as it is only used on the ephemeral EchoDeployment that's created before each test run, and destroyed right after.

Comment (zrdaley) [reply test/e2e/framework/framework.go]: I get what you mean; default it sort of arbitrary. I think it would be better to deepcopy the config map within the test as a local variable _before_ updating and then reset that way.

Comment (zrdaley) [reply test/e2e/framework/framework.go]: Another solution could be to have a `AddNginxConfigMapAnnotation` function that would add/update the annotation AND have a `RemoveNginxConfigMapAnnotation`. That way no copying would be needed and annotations could easily be added and revoked as needed.

Comment (zrdaley) [reply test/e2e/framework/framework.go]: A similar pattern is used in dynamic_configuation.go for `enableDynamicConfiguration` and `disableDynamicConfiguration`

Comment (andrewlouis93) [reply test/e2e/framework/framework.go]: Using Add and `RemoveNginxConfigMapAnnotation` might not work as expected - since there might have been a pre-existing value that we've overridden.

> I think it would be better to deepcopy the config map within the test as a local variable before updating and then reset at the end of the test.

I think this approach is pretty solid, and I wound up doing exactly this over here: https://github.com/Shopify/ingress/pull/44/files#diff-8700c46a8379c603b5e41ccbaff6b71eR123 (lines 123 - 129)

and here: 

https://github.com/Shopify/ingress/pull/44/files#diff-8700c46a8379c603b5e41ccbaff6b71eR153

I realized that those few lines would be repeated all over when we'd want to rollback - and returning the old configmap would avoid the trouble of having to rollback appropriately based on whether there _was_ a pre-existing value - or if there wasn't one

File: test/e2e/framework/framework.go [hunk 3]
```diff
@@ -259,6 +259,45 @@ func (f *Framework) matchNginxConditions(name string, matcher func(cfg string) b
...
+
+	if config == nil {
+		return nil, fmt.Errorf("nginx-configuration configMap does not exist")
+	}
+
+	return config, err
+}
+
+// UpdateNginxConfigMap updates the nginx-configuration's Data field
+func (f *Framework) UpdateNginxConfigMapData(k, v string) (error) {
```
Comment (zrdaley) [inline test/e2e/framework/framework.go]: nitpick for consistancy: `(k string, v string)`

File: Makefile [hunk 4]
```diff
@@ -19,7 +19,7 @@ BUILDTAGS=
 
 # Use the 0.0 tag for testing, it shouldn't clobber any release builds
 TAG?=0.12.0
-REGISTRY?=quay.io/kubernetes-ingress-controller
+REGISTRY?=shopify
```
Comment (zrdaley) [inline Makefile]: This should be changed in a separate PR/commit.

**Comment by zrdaley** [comment]
The configmap is updated in these tests and not reverted to the original state. 

Or should we leave as they were since we didn't write them?

File: test/e2e/framework/framework.go (hunk 3 above)
Comment (andrewlouis93) [reply test/e2e/framework/framework.go]: sounds good :)

**Comment by andrewlouis93** [comment]
@zrdaley you're right. There's an attempt to reset the configMap, since wherever the config is updated it seems to be in Before/After pairs - but there's an assumption being made about the initial state. Let's fix?

File: test/e2e/settings/no_auth_locations.go [hunk 5]
```diff
@@ -51,16 +52,24 @@ var _ = framework.IngressNginxDescribe("No Auth locations", func() {
...
 		Expect(s.ObjectMeta).NotTo(BeNil())
 
-		updateConfigmap(setting, noAuthPath, f.KubeClientSet)
-
 		bi := buildBasicAuthIngressWithSecondPath(host, f.Namespace.Name, s.Name, noAuthPath)
 		ing, err := f.EnsureIngress(bi)
 		Expect(err).NotTo(HaveOccurred())
 		Expect(ing).NotTo(BeNil())
+
+		if &defaultNginxConfigMapData == nil {
```
Comment (zrdaley) [inline test/e2e/settings/no_auth_locations.go]: Won't this always be true? I don't know if the if statement is necessary in all of these.

Comment (andrewlouis93) [reply test/e2e/settings/no_auth_locations.go]: Thought a bit more about this - I think it'd most likely always be false. The address of that stack variable isn't likely to be nil, since it's created as an empty map.

I think what I was thinking there was that the address would be nil until we assign a value to it - meaning it would only occur the first time - however I realize `defaultNginxConfigMapData` isn't a pointer.

Going to instead explicitly initialize the map to `nil` - should achieve what I was trying to without overcomplicating things :) Thanks Zenara!

File: test/e2e/framework/framework.go [hunk 6]
```diff
@@ -259,6 +259,80 @@ func (f *Framework) matchNginxConditions(name string, matcher func(cfg string) b
...
+	config, err := f.GetNginxConfigMapData()
+
+	if err != nil {
+		return err
+	}
+
+	config[key] = value
+
+	return f.SetNginxConfigMapData(config)
+}
```
Comment (ElvinEfendi) [inline test/e2e/framework/framework.go]: By implementing update config map this way you are now making two API calls. What shortcomings did https://github.com/Shopify/ingress/pull/45/files#diff-271db20be8cbad3b37eafe10f4eacbcfL115 have? It seemed super concise to me.
